Add `@stub` annotation to mark stubs

Duhemm opened the pull request:
By default, method marked with `@stub` will not link.

Fixes #971.

densh commented on nir/src/main/scala/scala/scalanative/nir/Attrs.scala:
```diff
@@ -29,12 +29,14 @@ object Attr {
   final case class PinAlways(dep: Global)           extends Pin
   final case class PinIf(dep: Global, cond: Global) extends Pin
   final case class PinWeak(dep: Global)             extends Pin
+  final case object Stub                            extends Attr
```
This one belongs next to Dyn in this file.

densh commented on nir/src/main/scala/scala/scalanative/nir/serialization/BinarySerializer.scala:
```diff
@@ -72,6 +72,8 @@ final class BinarySerializer(buffer: ByteBuffer) {
 
     case Attr.Dyn => putInt(T.DynAttr)
 
+    case Attr.Stub => putInt(T.StubAttr)
```
Lets put it in the same group as `Dyn`.

densh commented on tools/src/main/scala/scala/scalanative/linker/Linker.scala:
```diff
@@ -49,50 +49,52 @@ object Linker {
+                      weaks += global
+                  }
+
+              }
```
Please attach 5 runs of `tests/nativetest:nativeLinkNIR` (from a non-laptop machine if possible) starting from cold sbt start before and and after this change. We should be careful not to regress the linking performance.

densh commented on tools/src/main/scala/scala/scalanative/linker/Linker.scala:
```diff
@@ -49,50 +49,52 @@ object Linker {
-                    // comparing new dependencies with all signatures
-                    if (signatures(Global.genSignature(global))) {
+            load(workitem)
+              .filter(config.linkStubs || !_._4.attrs.isStub)
```
Lets move this to the success case inside fold with a comment for clarity.

densh commented:
Additionally: lets make sure that all `???` methods in javalib are marked as stubs.

densh commented:
One more thing: needs an update of `sbt.rst` to mention a new setting.

Duhemm commented:
@densh I've added all the `@stub`. `java.util.Formatter` depends on many things that were just stubbed out in `Calendar` and `TimeZone`, so I had to comment that code in `Formatter` and open issues to track it.

Where do the dependencies on `Class.getInterfaces` and `Class.getSuperclass` come from?

Regarding documenting the option, it looks `nativeConfig` is not mentioned at all (it is in `ScalaNativePluginInternal`). Should I introduce a new key, say `scalaNativeLinkStubs` in the sbt plugin?

densh commented:
We do need to expose an sbt setting. `nativeLinkStubs` sounds like a good fit for this.

Duhemm commented:
Before:
```
> tests/nativetest:nativeLink
[info] Linking (3700 ms)
[info] Discovered 4137 classes and 23843 methods
[info] Optimizing (4420 ms)
[info] Generating intermediate code (2683 ms)
[info] Produced 66 files
[info] Compiling to native code (1319 ms)
[info] Linking native code (327 ms)
[success] Total time: 15 s, completed Aug 31, 2017 2:05:30 PM
> tests/nativetest:nativeLink
[info] Linking (3423 ms)
[info] Discovered 4137 classes and 23843 methods
[info] Optimizing (3241 ms)
[info] Generating intermediate code (2074 ms)
[info] Produced 66 files
[info] Compiling to native code (1390 ms)
[info] Linking native code (338 ms)
[success] Total time: 11 s, completed Aug 31, 2017 2:05:43 PM
> tests/nativetest:nativeLink
[info] Linking (3632 ms)
[info] Discovered 4137 classes and 23843 methods
[info] Optimizing (3547 ms)
[info] Generating intermediate code (1716 ms)
[info] Produced 66 files
[info] Compiling to native code (1267 ms)
[info] Linking native code (340 ms)
[success] Total time: 11 s, completed Aug 31, 2017 2:05:57 PM
> tests/nativetest:nativeLink
[info] Linking (3248 ms)
[info] Discovered 4137 classes and 23843 methods
[info] Optimizing (3445 ms)
[info] Generating intermediate code (1552 ms)
[info] Produced 66 files
[info] Compiling to native code (1215 ms)
[info] Linking native code (341 ms)
[success] Total time: 14 s, completed Aug 31, 2017 2:06:12 PM
> tests/nativetest:nativeLink
[info] Linking (2999 ms)
[info] Discovered 4137 classes and 23843 methods
[info] Optimizing (3549 ms)
[info] Generating intermediate code (1524 ms)
[info] Produced 66 files
[info] Compiling to native code (1356 ms)
[info] Linking native code (350 ms)
[success] Total time: 10 s, completed Aug 31, 2017 2:06:24 PM
```

After:
```
> tests/nativetest:nativeLink
[info] Linking (3881 ms)
[info] Discovered 4038 classes and 23253 methods
[info] Optimizing (4170 ms)
[info] Generating intermediate code (2568 ms)
[info] Produced 66 files
[info] Compiling to native code (1116 ms)
[info] Linking native code (312 ms)
[success] Total time: 14 s, completed Aug 31, 2017 1:55:20 PM
> tests/nativetest:nativeLink
[info] Linking (3310 ms)
[info] Discovered 4038 classes and 23253 methods
[info] Optimizing (3053 ms)
[info] Generating intermediate code (1766 ms)
[info] Produced 66 files
[info] Compiling to native code (1153 ms)
[info] Linking native code (325 ms)
[success] Total time: 10 s, completed Aug 31, 2017 1:55:32 PM
> tests/nativetest:nativeLink
[info] Linking (3080 ms)
[info] Discovered 4038 classes and 23253 methods
[info] Optimizing (3008 ms)
[info] Generating intermediate code (1360 ms)
[info] Produced 66 files
[info] Compiling to native code (1162 ms)
[info] Linking native code (320 ms)
[success] Total time: 9 s, completed Aug 31, 2017 1:55:44 PM
> tests/nativetest:nativeLink
[info] Linking (3342 ms)
[info] Discovered 4038 classes and 23253 methods
[info] Optimizing (2908 ms)
[info] Generating intermediate code (1292 ms)
[info] Produced 66 files
[info] Compiling to native code (1142 ms)
[info] Linking native code (320 ms)
[success] Total time: 12 s, completed Aug 31, 2017 1:56:02 PM
> tests/nativetest:nativeLink
[info] Linking (2644 ms)
[info] Discovered 4038 classes and 23253 methods
[info] Optimizing (2920 ms)
[info] Generating intermediate code (1247 ms)
[info] Produced 66 files
[info] Compiling to native code (1100 ms)
[info] Linking native code (346 ms)
[success] Total time: 9 s, completed Aug 31, 2017 1:56:12 PM
```

Duhemm commented:
@densh I've added the timings, added the `@stub` annotation for everything that has been added post 0.3.2 and added a new setting key, `nativeLinkStub`.

densh commented on nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:
```diff
@@ -258,6 +258,7 @@ trait NirGenStat { self: NirGenPhase =>
             case ann if ann.symbol == NoInlineClass   => Attr.NoInline
             case ann if ann.symbol == InlineHintClass => Attr.InlineHint
             case ann if ann.symbol == InlineClass     => Attr.AlwaysInline
+            case ann if ann.symbol == StubClass       => Attr.Stub
```
Lets also support `@stub` on classes and modules, this should be trivial by adding similar line to `genClassAttrs`.

densh commented on javalib/src/main/scala/java/lang/ClassLoader.scala:
```diff
@@ -1,9 +1,15 @@
 package java.lang
 
+import scala.scalanative.native.stub
```
We typically omit `scala.` from the imports as `scalanative` is already specific enough.

densh commented on tools/src/test/scala/scala/scalanative/linker/StubSpec.scala:
```diff
@@ -0,0 +1,31 @@
+    }
+  }
+
+}
```
Needs a test for `@stub` classes and modules.

densh commented:
LGTM! 👍
